pkg/pillar/scripts/device-steps.sh

```diff
@@ -19,6 +19,9 @@ DPCDIR=$ZTMPDIR/DevicePortConfig
 FIRSTBOOTFILE=$ZTMPDIR/first-boot
 FIRSTBOOT=
 AGENTS="diag zedagent ledmanager nim nodeagent domainmgr loguploader tpmmgr vaultmgr zedmanager zedrouter downloader verifier baseosmgr wstunnelclient volumemgr watcher zfsmanager usbmanager"
+if grep -q kubevirt /run/eve-hv-type; then
```

Comment:
Could we just add `zedkube`, and the zedkube service itself would recognize, "oh no, I do not run under kubevirt, so I will do the `touch` and exit"?

I am concerned with polluting this with more specific logic. Everything else in `device-steps.sh` is generic. We have so many if/then and properties all over the place, keeping this clean would be nice.

Reply:
Yes, we could do that in `pkg/pillar/cmd/zedkube/nokube.go`. But note that we would have to update touch file periodically and could not just exit (otherwise watchdog will get triggered). Or we [do this](https://github.com/lf-edge/eve/blob/master/pkg/pillar/scripts/device-steps.sh#L233-L236), but then we just have another if/then.

Reply:
I didn't think about that. It isn't just the `wait_for_touch`, but the ongoing watchdog, isn't it? What happens with `diag` and watchdog? 

> Or we [do this](https://github.com/lf-edge/eve/blob/master/pkg/pillar/scripts/device-steps.sh#L233-L236), but then we just have another if/then

Better in one place than scattered about. The best solution would be to update this generically, so that we have some way of knowing without if/then scattered everywhere

Reply:
Yes, below `wait_for_touch` there is `touch "$WATCHDOG_FILE/$AGENT.touch"`, which enables periodic watchdog monitoring. 
I actually do not know why we do not enable watchdog for `diag`. Probably because it is not a critical microservice so we do not care if it will get stuck.

Reply:
But that makes it different than `zedkube`, which we do care about in watchdog.

I know it is "just one more if/then", but all of our areas across eve where we have too many of them and too much complexity started that way. It would be good to have a cleaner solution.

Reply:
I don't think it would be hard to make diag use a touch file. We don't have one because it isn't a critical service but if it makes things more uniform we can add it.

Reply:
But if we want to avoid this conditional we'd have to build a conditionally empty zedkube service which would just to the touch/keeprunning calls. I don't know if that makes anything more clear.

Reply:
Agreed. Which is why I think we need a more generic solution. Not "just one more if/then" custom case

Reply:
I can put a generic loop to touch function in some eve utility/library place, then have the zedkube/nokube.go call this function, and remove this if check in the device-steps.sh. will this be good enough?

Reply:
I don't understand

Reply:
Sorry maybe I didn't think though this.

Reply:
I think the way we use device-steps.sh is we don't have to add the runtime resource into a non-kubevirt eve operation.

Reply:
@deitch i was say that, no checks on device-steps.sh, always include zedkube. and in zedkube/nokube.go (since the go:build checks on non-kubevirt tag):
```

//go:build !kubevirt

package zedkube

...

type zedkubeContext struct {
	agentbase.AgentBase
}

// Run in this file is just stub for non-kubevirt hypervisors.
func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, arguments []string) int {
	logger := loggerArg
	log := logArg

	zedkubeCtx := zedkubeContext{}
	agentbase.Init(&zedkubeCtx, logger, log, agentName,
		agentbase.WithPidFile(),
		agentbase.WithWatchdog(ps, warningTime, errorTime),
		agentbase.WithArguments(arguments))

	stillRunning := time.NewTicker(stillRunningInterval)
	for {
		select {
		case <-stillRunning.C:
		}
		ps.StillRunning(agentName, warningTime, errorTime)
	}
	//panic("zedkube microservice is not built")
}
```

Reply:
> //go:build !kubevirt

This is the file that is built when not kubevirt variant? So this will turn the service into a perpetual loop?

Does this mean that all services always start and run. However, their behavior may vary:

* kubevirt: normal service running
* non-kubevirt: just an infinite loop, so that the service runs, but nothing happens

So from device-steps, it just looks like, "I have a list of services, I start them all, I run them all, I watch them all."

And since these all are threads, there is no overhead from launching an additional process with go runtime overhead, right? 

That's quite nice and elegant.

Reply:
The question though is if we need to introduce a new microservice in the first place.
We are already enhancing the existing microservices, like zedrouter, volumemgr, zedmanager and domainmgr, to handle kubevirt-specific workflow. So why to put these few things into a separate new microservice? Will there be more applications for zedkube in the cluster scenario?

Reply:
That is a _much_ better question than I asked. Do we need an additional µservice? I worked under the assumption that we do. If we do, then this approach is good. But best to answer Milan's question first.

Reply:
> And since these all are threads, there is no overhead from launching an additional process with go runtime overhead, right? 

The only overhead is the thread running and touching the file about every 15 seconds.

Reply:
> The question though is if we need to introduce a new microservice in the first place. We are already enhancing the existing microservices, like zedrouter, volumemgr, zedmanager and domainmgr, to handle kubevirt-specific workflow. So why to put these few things into a separate new microservice? Will there be more applications for zedkube in the cluster scenario?

There should be more things to come for the new service in cluster, we'll have it the publish like EdgeNodeClusterStatus, etc. Even the provision of cluster prefix can be done in that.